### PR TITLE
Add Index2d

### PR DESCRIPTION
**Pull request by piiswrong**
@tqchen

### CODE REVIEW

File: src/operator/mxnet_op.h [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    OP::Map(i, args...);
+  }
+}
+
+template<typename OP>
+struct Kernel<OP, gpu> {
+  template<typename ...Args>
+  inline static void Launch(mshadow::Stream<gpu> *s, int N, Args... args) {
+    using namespace mshadow::cuda;
+    int ngrid = std::min(kMaxGridNum, (N + kBaseThreadNum - 1) / kBaseThreadNum);
```
Comment (tqchen) [inline src/operator/mxnet_op.h]: if it is CUDA specific, consider put it in cu file

**Comment by tqchen** [comment]
please fix my other comment, and make sure the test pass before merge

**Comment by tqchen** [comment]
Another note is we want to be careful about naming, look out for existing names in say numpy and come up with an intuitive name,

**Comment by piiswrong** [comment]
array indexing is always implicit in numpy.
We can add this to getitem

**Comment by tqchen** [comment]
index2d seems to be not a very precise name for this. Maybe consider declare as a internal function, and overload getitem instead

**Comment by piiswrong** [comment]
I'll overload ndarray getitem. but index2d needs to be explicit because getitem doesn't work for symbolic

**Comment by tqchen** [comment]
any reason why cann't we overload getitem in symbolic case?

**Comment by piiswrong** [comment]
Because it's used for indexing out put
Overloading for indexing array is confusing
Tianqi Chen <notifications@github.com>于2017年1月18日 周三下午6:09写道：

**Comment by piiswrong** [comment]
I'll merge this first.
We need more discussion on whether we want to support advanced slicing for getitem. There are potential confusions from slicing by value vs by reference.

**Comment by sxjscience** [comment]
Would "take_index2d" be a better name? I also find index2d is not very precise.

File: src/operator/tensor/indexing_op.h [hunk 2]
```diff
@@ -327,6 +328,51 @@ void TakeOpBackward(const nnvm::NodeAttrs& attrs,
...
+inline bool BatchTakeOpType(const nnvm::NodeAttrs& attrs,
+                          std::vector<int> *in_attrs,
+                          std::vector<int> *out_attrs) {
+  CHECK_EQ(in_attrs->size(), 2);
+  if ((*in_attrs)[0] != -1) {
+    TYPE_ASSIGN_CHECK(*out_attrs, 0, (*in_attrs)[0]);
+  } else if ((*out_attrs)[0] != -1) {
+    TYPE_ASSIGN_CHECK(*in_attrs, 0, (*out_attrs)[0]);
+  }
+  TYPE_ASSIGN_CHECK(*in_attrs, 1, mshadow::kInt32);
```
Comment (mli) [inline src/operator/tensor/indexing_op.h]: @piiswrong `batch_take` seems requires the indices to be an int array, while `take` is ok for float array.

**Pull request merged**